src/cpp/src/llm_pipeline.cpp

```diff
@@ -0,0 +1,353 @@
+
+    std::string generate(std::string text, OptionalGenerationConfig generation_config, OptionalStreamerVariant streamer);
+    DecodedResults generate(std::vector<std::string> texts, OptionalGenerationConfig generation_config);
+    EncodedResults generate(ov::Tensor input_ids, std::optional<ov::Tensor> attention_mask, OptionalGenerationConfig generation_config, OptionalStreamerVariant streamer);
+
+    std::string apply_chat_template(std::string prompt, std::string role = "user") const;
+};
+
+}  // namespace genai
+}  // namespace ov
+
+using namespace std;
+
+
+ov::genai::LLMPipeline::LLMPipeline(
+    const std::string& model_path,
+    const ov::genai::Tokenizer& tokenizer,
+    const std::string& device,
+    const ov::AnyMap& plugin_config
+) {
+    m_pimpl = make_unique<LLMPipelineImpl>(model_path, tokenizer, device, plugin_config);
+}
+
+ov::genai::LLMPipeline::LLMPipelineImpl::LLMPipelineImpl(
+    const std::string& model_path,
+    const ov::genai::Tokenizer& tokenizer,
+    const std::string& device,
+    const ov::AnyMap& plugin_config
+): m_tokenizer(tokenizer) {
+    ov::Core core;
+    
+    std::string full_path = model_path;
+    if (!ov::genai::utils::is_xml(full_path))
+        full_path += "/openvino_model.xml";
+    try {
+        m_model_runner = core.compile_model(full_path, device, plugin_config).create_infer_request();
+    } catch (...) {
+        OPENVINO_THROW("Cannot compile_model from path " + full_path);
+    }
+}
+
+ov::genai::LLMPipeline::LLMPipeline(
+    const std::string& path, 
+    const std::string& device, 
+    const ov::AnyMap& config, 
+    const std::string& ov_tokenizers_path
+) {
+    m_pimpl = make_unique<LLMPipelineImpl>(path, device, config, ov_tokenizers_path);
+}
+
+ov::genai::LLMPipeline::LLMPipelineImpl::LLMPipelineImpl(
+    const std::string& path, 
+    const std::string& device, 
+    const ov::AnyMap& config, 
+    const std::string& ov_tokenizers_path
+): 
+    m_model_runner{ov::Core{}.compile_model(path + "/openvino_model.xml", device, config).create_infer_request()}, 
+    m_tokenizer{Tokenizer(path, device, ov_tokenizers_path)},
+    m_generation_config{from_config_json_if_exists(path)},
+    m_chat_template{from_tokenizer_json_if_exists(path)}
+ {}
+
+ov::genai::GenerationConfig ov::genai::LLMPipeline::LLMPipelineImpl::generation_config() const {
+    return m_generation_config;
+}
+
+ov::genai::GenerationConfig ov::genai::LLMPipeline::get_generation_config() const {
+    return m_pimpl->generation_config();
+}
+
+std::string ov::genai::LLMPipeline::LLMPipelineImpl::generate(
+    std::string text, 
+    OptionalGenerationConfig generation_config,
+    OptionalStreamerVariant streamer
+) {
+    GenerationConfig config = (generation_config.has_value()) ? *generation_config : m_generation_config;
+
+    if (is_chat_conversation) {
+        text = apply_chat_template(text);
+    }
+    auto kv_cache_len = m_model_runner.query_state()[0].get_state().get_shape()[2];
```

Comment:
it will download all cache from GPU to CPU, don't use this method directly. 
Try to manage cache length somehow else

Reply:
what are the alternatives? How can i get cache len?

Reply:
Maintain it manually as field of pipeline  based on prompt lengths

Reply:
removed, since i only need to know either it's >= 1 bool flag was enought